qiskit/primitives/base_estimator.py

```diff
@@ -155,16 +154,16 @@ def __init__(
         """
         if isinstance(circuits, QuantumCircuit):
             circuits = (circuits,)
-        self._circuits = tuple(circuits)
+        self._circuits = () if circuits is None else tuple(circuits)
 
         if isinstance(observables, SparsePauliOp):
             observables = (observables,)
-        self._observables = tuple(observables)
+        self._observables = () if observables is None else tuple(observables)
 
         # To guarantee that they exist as instance variable.
         # With only dynamic set, the python will not know if the attribute exists or not.
-        self._circuit_ids = self._circuit_ids
-        self._observable_ids = self._observable_ids
+        self._circuit_ids: list[int] = self._circuit_ids
```

Comment:
If we get rid of passing in indexes to `run()`, then we won't even need this `circuit_ids` in the base class, right? The default implementation doesn't even need to do data caching since it's all local.

Reply:
Exactly right. My only concern is the performance of `qiskit-ibm-runtime` version, so if runtime does not require integer indices for the performance, I want to get rid of it.

Reply:
Yeah I don't think we'll be using integer indices because it's confusing to the user and hard to keep track if the user creates multiple instances of `sampler` etc. 

The plan now is to build a dictionary indexed by the object id, e.g.
`circuit_dict = {id(qc1): qc1, id(qc2): qc2}`

Then we can find the right circuit quickly and also pass just the object ids to the server once they're cached.

Reply:
`circuit_ids` are only used in deprecated methods now. So it will be removed in the future.

Reply:
I see `circuit_ids` is still used in the `_run()` methods?

Reply:
In the reference implementation, yes. I can remove it when we remove `__call__`.